module-apps/apps-common/popups/presenter/WallpaperPresenter.cpp

```diff
@@ -2,58 +2,91 @@
-                                              std::shared_ptr<WallpaperBase> quoteWallpaper,
-                                              std::shared_ptr<WallpaperBase> logoWallpaper)
+    void WallpaperPresenter::setupWallpaper(Item *parent)
     {
-        this->clockWallpaper = clockWallpaper;
-        this->quoteWallpaper = quoteWallpaper;
-        this->logoWallpaper  = logoWallpaper;
-    }
-
-    void WallpaperPresenter::setupWallpaper()
-    {
-        buildWallpapers();
+        buildWallpapers(parent);
 
         auto request = std::make_unique<app::manager::GetWallpaperOptionRequest>();
         auto task    = app::AsyncRequest::createFromMessage(std::move(request), service::name::appmgr);
         auto cb      = [&](auto response) {
             auto result = dynamic_cast<app::manager::GetWallpaperOptionResponse *>(response);
-            switch (result->getWallpaperOption()) {
-            case WallpaperOption::Clock:
-                clockWallpaper->show();
-                quoteWallpaper->hide();
-                logoWallpaper->hide();
-                break;
-            case WallpaperOption::Quote:
-                clockWallpaper->hide();
-                quoteWallpaper->show();
-                logoWallpaper->hide();
-                break;
-            case WallpaperOption::Logo:
-                clockWallpaper->hide();
-                quoteWallpaper->hide();
-                logoWallpaper->show();
-                break;
+            if (!clockWallpaperForced) {
+                switchWallpaper(result->getWallpaperOption());
             }
+            selectedOption = result->getWallpaperOption();
             return true;
         };
         task->execute(application, this, cb);
     }
 
-    void WallpaperPresenter::buildWallpapers()
+    void WallpaperPresenter::switchWallpaper(WallpaperOption option)
+    {
+        switch (option) {
+        case WallpaperOption::Clock:
+            notificationsModel->attachPresenter(clockWallpaper->getNotificationsPresenter());
+            clockWallpaper->show();
+            quoteWallpaper->hide();
+            logoWallpaper->hide();
+            break;
+        case WallpaperOption::Quote:
+            notificationsModel->attachPresenter(quoteWallpaper->getNotificationsPresenter());
+            clockWallpaper->hide();
+            quoteWallpaper->show();
+            logoWallpaper->hide();
+            break;
+        case WallpaperOption::Logo:
+            notificationsModel->attachPresenter(logoWallpaper->getNotificationsPresenter());
+            clockWallpaper->hide();
+            quoteWallpaper->hide();
+            logoWallpaper->show();
+            break;
+        }
+    }
+
+    void WallpaperPresenter::buildWallpapers(Item *parent)
     {
-        clockWallpaper->build();
-        quoteWallpaper->build();
-        logoWallpaper->build();
+        clockWallpaper = std::make_shared<WallpaperClock>(parent);
```

Comment:
i would recommend moving that to WallpaperModel -> and just access here. Presenter is not best place to hold widgets in my opinion. It should only control them.